misc/allow-block-list/src/lib.rs

```diff
@@ -0,0 +1,429 @@
+// Copyright 2023 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use libp2p_core::{Endpoint, Multiaddr};
+use libp2p_identity::PeerId;
+use libp2p_swarm::{
+    dummy, CloseConnection, ConnectionDenied, ConnectionId, FromSwarm, NetworkBehaviour,
+    NetworkBehaviourAction, PollParameters, THandler, THandlerInEvent, THandlerOutEvent,
+};
+use std::collections::{HashSet, VecDeque};
+use std::fmt;
+use std::task::{Context, Poll, Waker};
+use void::Void;
+
+/// A [`NetworkBehaviour`] that can act as an allow or block list for peers.
+#[derive(Debug)]
+pub struct Behaviour<S> {
```

Comment:
The generic parameter `S` allows for a lot of code de-duplication below at the expense of _some_ complexity for the user. How about adding two type aliases to make the users live easier?

``` rust
type Allow = Behaviour<Allow>;
type Block = Behaviour<Block>;
```

Reply:
It is not totally inline with the naming convention but I am happy to add it :)

Reply:
As I was writing the examples, I changed my mind on this. It is not in-line with the naming conventions so I opted to not add a type alias.

Reply:
:+1: resolving here.